infrastructure/src/main/java/org/corfudb/infrastructure/logreplication/replication/LogReplicationSourceManager.java

```diff
@@ -247,17 +226,75 @@ public LogReplicationEntry receive(LogReplicationEntry message) {
             logReplicationFSM.input(new LogReplicationEvent(LogReplicationEventType.LOG_ENTRY_SYNC_REPLICATED,
                 new LogReplicationEventMetadata(message.getMetadata().getSyncRequestId(), message.getMetadata().getTimestamp())));
         } else if (message.getMetadata().getMessageMetadataType() == MessageType.SNAPSHOT_REPLICATED) {
             log.debug("Snapshot sync ACK received on base timestamp {}", message.getMetadata().getSnapshotTimestamp());
+            lastAckedTimestamp.set(message.getMetadata().getTimestamp());
             logReplicationFSM.input(new LogReplicationEvent(LogReplicationEventType.SNAPSHOT_SYNC_COMPLETE,
                     new LogReplicationEventMetadata(message.getMetadata().getSyncRequestId(), message.getMetadata().getTimestamp(),
                             message.getMetadata().getTimestamp())));
         } else {
             log.debug("Received data message of type {} not an ACK", message.getMetadata().getMessageMetadataType());
         }
-
         return null;
     }
+
+    /**
+     * For the given replication runtime, query max stream tail for all streams to be replicated.
+     *
+     * @return max tail of all streams to be replicated for the given runtime
+     */
+    private long getMaxReplicatedStreamsTail() {
+        Map<UUID, Long> tailMap = runtime.getAddressSpaceView().getAllTails().getStreamTails();
+        long maxTail = Address.NON_ADDRESS;
+        for (String s : config.getStreamsToReplicate()) {
+            UUID streamUuid = CorfuRuntime.getStreamID(s);
+            if (tailMap.get(streamUuid) != null) {
+                long streamTail = tailMap.get(streamUuid);
+                maxTail = Math.max(maxTail, streamTail);
+            }
+        }
+        return maxTail;
+    }
+
+    /**
+     * Given a timestamp acked by the receiver, calculate how many entries remain to be sent for all replicated streams.
+     *
+     * @param ackedTimestamp Timestamp ack'd by the receiver
+     *
+     * For Log Entry Sync, this function returns the total number of entries remaining to be sent across all replicated
+     * streams.
+     * For Snapshot Sync, each entry sent is a snapshot.  So this function returns the total number of snapshots
+     * remaining to be sent.
+     * If the ack'd timestamp is uninitialized, it returns 100%, which means no replication has been done.
+     */
+    private long calculateRemainingEntriesToSend(long ackedTimestamp) {
+        long maxReplicatedStreamTail = getMaxReplicatedStreamsTail();
+
+        if (ackedTimestamp == Address.NON_ADDRESS) {
+            return FULL_REPLICATION_REMAINING_PERCENT;
+        }
+        long remainingEntriesToSend = 0;
+        for (String stream : config.getStreamsToReplicate()) {
+            UUID streamId = CorfuRuntime.getStreamID(stream);
+            StreamAddressRange range = new StreamAddressRange(streamId, maxReplicatedStreamTail, ackedTimestamp);
```

Comment:
I just want to clarify. ackedTimestamp is the last timestamp acked by the receiver that we receive when the entry is replicated, so it should be smaller than the maxReplicatedStreamTail, for the range to be valid right?